### PR TITLE
add sanity and out of bounds checks for keymaster TA serializers

### PR DESCRIPTION
**Pull request by vchong**
@ruchi393 Please review and approve. Thanks.

### CODE REVIEW

File: keymaster/ta/mbedtls_proxy.c [hunk 1]
```diff
@@ -272,8 +270,39 @@ static keymaster_error_t convert_epoch_to_date_str(uint32_t sec,
...
+		DMSG("Saturday");
+		break;
+	default:
+		break;
+	}
+	if (leap_year_ind) {
+		DMSG("%04u is a leap year", ntp_year);
+	} else {
+		DMSG("%04u is not a leap year", ntp_year);
+	}
```
Comment (ruchi393) [inline keymaster/ta/mbedtls_proxy.c:305]: Any specific reason for adding so many debug messages in this patch ?

File: keymaster/ta/keystore_ta.c [hunk 2]
```diff
@@ -485,6 +486,11 @@ static keymaster_error_t TA_importKey(TEE_Param params[TEE_NUM_PARAMS])
 	if (res != KM_ERROR_OK)
 		goto out;
 	TA_add_origin(&params_t, KM_ORIGIN_IMPORTED, true);
+	if (TA_is_out_of_bounds(in, in_end, sizeof(key_format))) {
```
Comment (ruchi393) [inline keymaster/ta/keystore_ta.c:619]: Add a bank line before this check

File: keymaster/ta/parsel.c [hunk 3]
```diff
@@ -145,9 +169,17 @@ static bool param_deserialize(keymaster_key_param_t *param, uint8_t **buf_ptr,
...
 	case KM_BIGNUM:
 	case KM_BYTES:
+		if (TA_is_out_of_bounds(*buf_ptr, end, sizeof(uint32_t))) {
+			EMSG("Out of input array bounds on deserialization");
+			return false;
+		}
 		TEE_MemMove(&param->key_param.blob.data_length, *buf_ptr,
 			    sizeof(uint32_t));
 		*buf_ptr += sizeof(uint32_t);
+		if (TA_is_out_of_bounds(*buf_ptr, end, sizeof(offset))) {
```
Comment (ruchi393) [inline keymaster/ta/parsel.c]: Should we combine this check with earlier TA_is_out_of_bounds check

something like
if (TA_is_out_of_bounds(*buf_ptr, end, sizeof(uint32_t) + sizeof(offset)))

File: keymaster/ipc/optee_keymaster_ipc.cpp [hunk 4]
```diff
@@ -251,7 +251,15 @@ keymaster_error_t optee_keymaster_call(uint32_t cmd,
...
-	return rsp->error;
+    } else if (res != KM_ERROR_OK) {
+	/*
+	 * rsp->error is KM_ERROR_OK but res isn't? The only time this happens
+	 * is in the case of param_types != exp_param_types in
+	 * TA_InvokeCommandEntryPoint() so res can't be serialized into rsp.
+	 */
+	ALOGE("Response of size %d contained error code %d\n", (int)rsp_size,
+	      (int)res);
+	return (keymaster_error_t)res;
```
Comment (ruchi393) [inline keymaster/ipc/optee_keymaster_ipc.cpp:267]: res looks like TEE_ERROR. Wouldn't typecasting it to keymaster_error_t result in wrong errorcode

File: keymaster/ipc/optee_keymaster_ipc.cpp [hunk 5]
```diff
@@ -253,9 +253,10 @@ keymaster_error_t optee_keymaster_call(uint32_t cmd,
 	      (int)rsp->error);
     } else if (res != KM_ERROR_OK) {
 	/*
-	 * rsp->error is KM_ERROR_OK but res isn't? The only time this happens
-	 * is in the case of param_types != exp_param_types in
-	 * TA_InvokeCommandEntryPoint() so res can't be serialized into rsp.
+	 * rsp->error is KM_ERROR_OK but res isn't? This happens when:
+	 * 1. param_types != exp_param_types in TA_InvokeCommandEntryPoint()
+	 * 2. req or rsp (in or out bufptr in keystore_ta.c) are NULL
+	 *    so res can't be serialized into rsp
```
Comment (ruchi393) [inline keymaster/ipc/optee_keymaster_ipc.cpp:259]: Looks like this change doesn't belong to this commit

File: keymaster/ta/keystore_ta.c [hunk 6]
```diff
@@ -147,12 +147,16 @@ static keymaster_error_t TA_configure(TEE_Param params[TEE_NUM_PARAMS])
...
+	uint8_t *out_end = NULL;
+	size_t out_size = 0;
 	keymaster_error_t res = KM_ERROR_OK;
 
 	in = (uint8_t *)params[0].memref.buffer;
 	in_size = (size_t)params[0].memref.size;
 	in_end = in + in_size;
 	out = (uint8_t *)params[1].memref.buffer;
+	out_size = (size_t)params[1].memref.size; /* limited to 8192 */
+	out_end = out + out_size;
```
Comment (ruchi393) [inline keymaster/ta/keystore_ta.c:160]: Shouldn't we be also adding out of bound check for out ?

Comment (ruchi393) [reply keymaster/ta/keystore_ta.c:160]: I just re-read the commit...so you are going t do that in next commit. Ignore this comment.

**Comment by ruchi393** [comment]
Overall it LGTM. Have added few queries/comments I had.

What are the test results with these patches ?

File: keymaster/ta/mbedtls_proxy.c (hunk 1 above)
Comment (vchong) [reply keymaster/ta/mbedtls_proxy.c:305]: This is wrt https://github.com/linaro-swg/kmgk/pull/21 61661f5ba870a346ff32fb3231cc8621f2c985c6. The warning 'variable set but not used' reappears when log level is < DBG, so use the variable itself to determine which debug logs to print.

File: keymaster/ta/keystore_ta.c (hunk 2 above)
Comment (vchong) [reply keymaster/ta/keystore_ta.c:619]: ok

File: keymaster/ta/parsel.c (hunk 3 above)
Comment (vchong) [reply keymaster/ta/parsel.c]: good idea, will do

File: keymaster/ipc/optee_keymaster_ipc.cpp (hunk 4 above)
Comment (vchong) [reply keymaster/ipc/optee_keymaster_ipc.cpp:267]: The `TEE_Result` from `TA_InvokeCommandEntryPoint()` is actually a `keymaster_error_t` since all the functions called by `TA_InvokeCommandEntryPoint()` return a `keymaster_error_t` so here we cast it to match.

File: keymaster/ipc/optee_keymaster_ipc.cpp (hunk 5 above)
Comment (vchong) [reply keymaster/ipc/optee_keymaster_ipc.cpp:259]: It does. In the keystore_ta.c checks below, if either the input or output buffer is null, we return immediately without writing the error code to the output buffer, and land here in the CA, so a comment is added to note the case.

**Comment by vchong** [comment]
> Overall it LGTM. Have added few queries/comments I had.
> 
> What are the test results with these patches ?

All comments addressed and forced pushed since changes are minor.
Verified no regressions in test results with these patches on master.

**Comment by vchong** [comment]
Pushed an additional patch (`km: ta: add oob check to TA_fill_characteristics()`) missed in the original series.

**Comment by ruchi393** [comment]
LGTM 

`Acked-by: Ruchika Gupta <ruchika.gupta@linaro.org>`
